### PR TITLE
 arch/xtensa： Update I2C driver.

### PR DESCRIPTION
**Pull request by xiarihf**
1.  Enble 10 bits slave address in I2C driver.
2.  Fix overflow error in i2c_esp32_transmit.
     The timeout value is set in Microsecond, it may be a large value. Thus, priv->timeout * NS_COUNT_IN_MS may overflow.
3.  Fix I2C reading larg data issue.
      a. Modify the I2C reading interfaces for larg data reading.
         The instructions of I2C reading in ESP32 datasheet has a mistake.
         It cause this issue.
      b. Add i2C bus busy status checking to wait for transmission complete.
4. Modify I2C default SDA pin. 
    Pin IO5 is used in LCD. Thus, I2C SDA is changed to pin IO2 in wrover kit board.

Signed-off-by: xiarihf qiang3.zhang@samsung.com

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [07516208d53e54c90e55587032ce939563b73da9] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [07516208d53e54c90e55587032ce939563b73da9] - Code Rule Check OK.

File: build/configs/esp_wrover_kit/tc/defconfig [hunk 1]
```diff
@@ -95,7 +95,7 @@ CONFIG_ESP32_I2C0_MODE_MASTER=y
 # CONFIG_ESP32_I2C0_MODE_SLAVE is not set
 CONFIG_ESP32_I2C0_SCLK_PIN=4
 # CONFIG_ESP32_I2C0_SCLK_PULLUP_EN is not set
-CONFIG_ESP32_I2C0_SDA_PIN=5
+CONFIG_ESP32_I2C0_SDA_PIN=2
```
Comment (an4967) [inline build/configs/esp_wrover_kit/tc/defconfig]: Why did you change the default SDA number?

File: os/arch/xtensa/src/esp32/esp32_i2c.c [hunk 2]
```diff
@@ -670,28 +702,47 @@ static void i2c_esp32_send_write_cmd(struct esp32_i2c_priv_s *priv, int addr_byt
...
-static void i2c_esp32_send_read_cmd(struct esp32_i2c_priv_s *priv, int addr_bytes, int cmd_start_index)
+static int i2c_esp32_send_read_cmd(struct esp32_i2c_priv_s *priv, int addr_bytes, int cmd_start_index)
 {
 	struct i2c_msg_s *msg = priv->transfer_msg;
 	volatile struct i2c_esp32_cmd *cmd = (void *)I2C_COMD0_REG(priv->i2c_num);
 	cmd += cmd_start_index;
 
-	uint32_t to_read = min(I2C_ESP32_BUFFER_SIZE, msg->length - 1);
-	volatile struct i2c_esp32_cmd *wait_cmd = NULL;
+	if (msg->length <= 0){
```
Comment (ChinaTizenRT) [inline os/arch/xtensa/src/esp32/esp32_i2c.c]: space needed

File: os/arch/xtensa/src/esp32/esp32_i2c.c [hunk 3]
```diff
@@ -332,23 +351,33 @@ static int i2c_esp32_isr(int irq, FAR void *context, FAR void *arg)
...
 				if (priv->status == I2C_STATUS_READ) {
 					i2c_esp32_get_i2c_data(priv);
-					i2c_esp32_send_read_cmd(priv, 0, 0);
+					ret = i2c_esp32_send_read_cmd(priv, 0, 0);
 				} else if (priv->status == I2C_STATUS_WRITE) {
 					i2c_esp32_reset_txfifo(priv->i2c_num);
-					i2c_esp32_send_write_cmd(priv, 0, 0);
+					ret = i2c_esp32_send_write_cmd(priv, 0, 0);
+				}
+				if (ret >= 0){
```
Comment (ChinaTizenRT) [inline os/arch/xtensa/src/esp32/esp32_i2c.c]: space needed

File: os/arch/xtensa/src/esp32/esp32_i2c.c [hunk 4]
```diff
@@ -332,23 +351,33 @@ static int i2c_esp32_isr(int irq, FAR void *context, FAR void *arg)
 				I2C[priv->i2c_num]->int_ena.time_out = 0;
 				I2C[priv->i2c_num]->int_clr.time_out = 1;
 				priv->status = I2C_STATUS_TIMEOUT;
+				ets_printf("status is %d, I2C_TIME_OUT_INT_ST_M!\n", status);
```
Comment (ChinaTizenRT) [inline os/arch/xtensa/src/esp32/esp32_i2c.c]: this log need keep it?

File: os/arch/xtensa/src/esp32/esp32_i2c.c [hunk 5]
```diff
@@ -144,6 +150,7 @@ struct esp32_i2c_priv_s {
 
 	int xfer_speed;
 	uint32_t slave_addr;
+	uint8_t slave_addr_bits;
```
Comment (ChinaTizenRT) [inline os/arch/xtensa/src/esp32/esp32_i2c.c]: 10bits slave device access feature  is not related with  debuging long data reading fail issue,
spliting it one more PR may be clear.

File: os/arch/xtensa/src/esp32/esp32_i2c.c [hunk 6]
```diff
@@ -646,21 +678,21 @@ static void i2c_esp32_get_i2c_data(struct esp32_i2c_priv_s *priv)
...
+	uint32_t i, to_send = 0;
 
 	volatile struct i2c_esp32_cmd *cmd = (void *)I2C_COMD0_REG(priv->i2c_num);
 	cmd += cmd_start_index;
 
 	if (msg->length == 0) {
 		*cmd = (struct i2c_esp32_cmd) {
 			.opcode = I2C_ESP32_OP_STOP,
-			.ack_en = false,
+			.ack_en = 0,
```
Comment (an4967) [inline os/arch/xtensa/src/esp32/esp32_i2c.c]: Isn't it better using 'false' in readability perspective?

File: os/arch/xtensa/src/esp32/esp32_i2c.c [hunk 7]
```diff
@@ -706,36 +757,30 @@ static void i2c_esp32_send_read_cmd(struct esp32_i2c_priv_s *priv, int addr_byte
...
 
-	priv->last_data_count = to_read;
+	*cmd++ = (struct i2c_esp32_cmd) {
+		.opcode = I2C_ESP32_OP_END
+	};
+
+	I2C[priv->i2c_num]->int_ena.end_detect = 1;
 
+	priv->last_data_count = to_read;
+	return to_read;
```
Comment (an4967) [inline os/arch/xtensa/src/esp32/esp32_i2c.c]: For this, how about change return type of 'i2c_esp32_send_read_cmd' to 'uint32_t' ?

**Comment by an4967** [comment]
I tested this PR, and it worked very well.

File: os/arch/xtensa/src/esp32/esp32_i2c.c (hunk 2 above)
Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: fixed.

File: os/arch/xtensa/src/esp32/esp32_i2c.c (hunk 3 above)
Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: fixed

File: os/arch/xtensa/src/esp32/esp32_i2c.c (hunk 4 above)
Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: removed.

File: os/arch/xtensa/src/esp32/esp32_i2c.c (hunk 5 above)
Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: yes, thanks.

File: build/configs/esp_wrover_kit/tc/defconfig (hunk 1 above)
Comment (xiarihf) [reply build/configs/esp_wrover_kit/tc/defconfig]: Pin IO5 is used in LCD. Thus, I2C SDA is changed to pin IO2 in wrover kit board.

File: os/arch/xtensa/src/esp32/esp32_i2c.c (hunk 6 above)
Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: ack_en is uint32_t, it's not a good choise to use bool value.

File: os/arch/xtensa/src/esp32/esp32_i2c.c (hunk 7 above)
Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: It will return "ESP32_I2C_READ_ERROR", which is " -1" for error code.

**Comment by seinfra** [comment]
Target : [cdcc703c282099e62dd01ca921cc9a20aa6d68d1] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [cdcc703c282099e62dd01ca921cc9a20aa6d68d1] - Code Rule Check (C++) OK.

**Comment by xiarihf** [comment]
> At the end of Line:223, a comma(,) is needed. If CONFIG_ESP32_I2C0_SDA_PIN is not set, there would be a build error.

oh, yes, thanks. I need to create a new PR.

File: os/arch/xtensa/src/esp32/esp32_i2c.c [hunk 8]
```diff
@@ -646,21 +677,21 @@ static void i2c_esp32_get_i2c_data(struct esp32_i2c_priv_s *priv)
 0:  no slave address
 1: 7bit slave address
 2: 10bit slave address*/
-static void i2c_esp32_send_write_cmd(struct esp32_i2c_priv_s *priv, int addr_bytes, int cmd_start_index)
+static int i2c_esp32_send_write_cmd(struct esp32_i2c_priv_s *priv, int addr_bytes, int cmd_start_index)
 {
 	struct i2c_msg_s *msg = priv->transfer_msg;
-	uint32_t i = 0;
+	uint32_t i, to_send = 0;
```
Comment (NilabhShant) [inline os/arch/xtensa/src/esp32/esp32_i2c.c]: please declare new variable in new line.

like: 
           uint32_t i;
           uint32_t to_send = 0;

**Comment by seinfra** [comment]
Target : [7e7fde32d3fa0fdc1781ffb731845f41d8088a62] - Code Rule Check Result:
os/arch/xtensa/src/esp32/esp32_i2c.c:326: ERROR: [SPC_M_KWD] space required before the open brace '{'

File: os/arch/xtensa/src/esp32/esp32_i2c.c [hunk 9]
```diff
@@ -670,28 +701,47 @@ static void i2c_esp32_send_write_cmd(struct esp32_i2c_priv_s *priv, int addr_byt
...
+			*cmd = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_STOP
+			};
+			return OK;
+		}
+
+		return ESP32_I2C_READ_ERROR;
+	}
+
+	int length = msg->length;
```
Comment (NilabhShant) [inline os/arch/xtensa/src/esp32/esp32_i2c.c]: Is it fine to declare a variable like this?

**Comment by seinfra** [comment]
Target : [7e7fde32d3fa0fdc1781ffb731845f41d8088a62] - Code Rule Check (C++) OK.

File: os/arch/xtensa/src/esp32/esp32_i2c.c (hunk 8 above)
Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: Fixed, thanks.

File: os/arch/xtensa/src/esp32/esp32_i2c.c (hunk 9 above)
Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: yes, fixed. thanks.

**Comment by seinfra** [comment]
Target : [6581eacc358001a937e4d6520d629d812e768593] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [6581eacc358001a937e4d6520d629d812e768593] - Code Rule Check (C++) OK.

File: os/arch/xtensa/src/esp32/esp32_i2c.c [hunk 10]
```diff
@@ -313,12 +320,24 @@ static int i2c_get_timeout(i2c_port_t i2c_num, int* timeout)
 	return 0;
 }
 
+static int i2c_esp32_check_bus(struct esp32_i2c_priv_s *priv)
```
Comment (Taejun-Kwon) [inline os/arch/xtensa/src/esp32/esp32_i2c.c]: I think this is pretty risky.
You'd better  clock instead of loop count, because if there are many threads, running time will be increased.
Also naming should be changed, how about i2c_esp32_wait_until_ready or something.

File: os/arch/xtensa/src/esp32/esp32_i2c.c [hunk 11]
```diff
@@ -335,20 +354,31 @@ static int i2c_esp32_isr(int irq, FAR void *context, FAR void *arg)
 				sem_post(&priv->sem_isr);
 				break;
 			} else if (status & I2C_END_DETECT_INT_ST_M) {
+				I2C[priv->i2c_num]->int_ena.end_detect = 0;
 				I2C[priv->i2c_num]->int_clr.end_detect = 1;
+				if (i2c_esp32_check_bus(priv) == ESP32_BUS_BUSY) {
+					priv->status = I2C_STATUS_TIMEOUT;
```
Comment (Taejun-Kwon) [inline os/arch/xtensa/src/esp32/esp32_i2c.c]: This case, it failed because of i2c state is busy but returns OK, is it correct flow? I don't think so.

Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: it will use priv->status to notice i2c_esp32_transmit, it will recieve I2C_STATUS_TIMEOUT then returns  -ETIMEDOUT. please check it.

File: os/arch/xtensa/src/esp32/esp32_i2c.c (hunk 10 above)
Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: Hi, Mr. Kwon, This is interrupt handler, we could not use delay or sleep funcion, and also, there is no such multiple thread cases, because interrupt handler is unique, it will be called only when interrupt coming out.

Comment (Taejun-Kwon) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: Got it.

File: os/arch/xtensa/src/esp32/esp32_i2c.c (hunk 11 above)
Comment (sangwon03) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: @xiarihf 
you mean that is it correct to setup hsi2c although i2c state is busy?

Comment (xiarihf) [reply os/arch/xtensa/src/esp32/esp32_i2c.c]: Sorry, i did not catch your question. do you mean hsi2c_setup function? it is only called in driver initializing in booting period.  The i2c bus is transmitting data, and the status reg bit I2C_BUS_BUSY is 1, after it transmitting data, it will be 0.
